nevergrad/functions/games/battleship.py

```diff
@@ -0,0 +1,337 @@
+            list_of_x += [(i, j) for j, x in enumerate(b) if x == 'x']
+
+        while True:
+            if verbose:
+                print("Creating an empty board.")
+            candidate = battleship(version=self.version)
+            failed = False
+            available_ships = [x_ for x_ in self.sizes()]
+
+            # First, put ships in order to cover the 'x' and not the '*'.
+            random.shuffle(available_ships)
+            random.shuffle(list_of_x)
+            for x in list_of_x:
+                if verbose:
+                    print("we must cover " + str(x))
+                h = x[0]
+                v = x[1]
+                if candidate.board[h][v] == 'o':  # It's ok, there is already a ship here, we don't have to cover it.
+                    if verbose:
+                        print("already covered, cool!")
+                    continue
+                # Let us pick up the next ship.
+                if not available_ships:
+                    if verbose:
+                        print("no more ships!")
+                    failed = True
+                    break
+                ship = available_ships[-1]
+
+                # Randomly draw its direction and shift, so that it covers (h,v).
+                while True:
+                    if verbose:
+                        print("trying to put it on the board...", h, v, ship)
+                    shift = random.randint(1-ship, 0)
+                    if random.choice([True, False]):
+                        dx = 1
+                        dy = 0
+                    else:
+                        dx = 0
+                        dy = 1
+                    if h + shift*dx >= 0 and h+(ship+shift)*dx <= N and v+shift*dy >= 0 and v+(ship+shift)*dy <= N:
+                        break
+
+                if verbose:
+                    print("ship found, with its position.")
+                # Put this ship!
+                for a in range(shift, shift+ship):
+                    if board[h+a*dx][v+a*dy] == '*' or candidate.board[h+a*dx][v+a*dy] == 'o':
+                        failed = True  # There is no ship here! or we already put one.
+                        break
+                if failed:
+                    break  # Restart from scratch!
+                for a in range(shift, shift+ship):
+                    candidate.board[h+a*dx][v+a*dy] = 'o'
+                del available_ships[-1]
+
+            if failed:  # We failed to cover this x!
+                continue
+
+            if verbose:
+                print("Intermediate board with known ships covered:")
+                print(candidate)
+                print(" ... should cover:")
+                print(self)
+                print("Available ships:", available_ships)
+
+            # Now let us put the remaining ships!
+            while available_ships and not failed:
+                if verbose:
+                    print("Remaining ships :" + str(available_ships))
```

Comment:
Using fstring is more readable `print("Remaining ships : {available_ships}")`

Reply:
a f is missing presumably, and ok done!